review hydromt plugin cookiecutter

hboisgon opened the pull request:
Review of the cookiecutter hydromt plugin boilerplate

hboisgon committed:
review hydromt plugin cookiecutter

hboisgon commented:
Hi @savente93.
Love your cookiecutter solution to quickly set up a nice hydromt plugin! Great job :)

I changed some things here and there and added a couple more template files.

Some questions I have:

1. why did you use Quarto and md files for the docs rather than hydromt templates?
2. I think we should still clean the few docs there is, especially python section and decide if we include the main hydromt headers or leave it completely blank/open
3. I added dependencies used in the template files to pyproject.toml as well as extra dependencies needed for test/dev tools. I do not know Quarto so not sure what we should add for the doc dependencies. Was I correct to add these dependencies?
4. Did you also actually try and publish one plugin and checked if linting/tests/docs are correctly passed/built?

aslowwriter commented on {{cookiecutter.project_slug}}/pyproject.toml:
```diff
@@ -20,21 +25,33 @@ classifiers = [
+dev = [
+    "black",            # code formatting
+    "ruff",             # code linting
+    "mypy",             # static type checking
```
including mypy here is fine, but the template isn't curently steup to use it in anyway so the people making the plugin will have to use it manually. I also don't know if it will throw warnings or errors out of the box. Given that Hydromt's support for mypy is spotty at best the chance is fairly high, so not 100% yet on whether we want to include this.

hboisgon replied:
I do not it (was a Copilot suggestion) so I leave it to you to decide if we include it or not :)

aslowwriter commented on cookiecutter.json:
```diff
@@ -3,8 +3,9 @@
-  "model_name": "PluginModel", 
-  "main_file_basename": "{{ cookiecutter.model_name.lower().replace('model','') }}", 
+  "model_classname": "PluginModel",
+  "model_type": "Model", 
```
we can include this model_type part of the template but do note that the user can fill in anything they like here, so there will be no checking, and it's very possible to generate a broken template  if you make a typo here. If you're fine with that we can include this, but personally I'm hesitant.

hboisgon replied:
I think it helps to get a good start to start with the correct submodel class. Maybe I can include the list in the suggestions rather than just Model (Model / GridModel / MeshModel / LumpedModel)? Shame we can't do check with cookiecutter but in a way any typo in the cookiecutter argument needs to be checked so I still think it's nice to include

aslowwriter replied:
there is a mechanic in coockiecutter to be able to choose from options so that would help but currently you can only choose one. There is a feature request to be able to choose multiple options, but that is for now unmerged and I don't know if/when that will land. I think for now this is fine, and we can always revisit it later when that feature is published.

aslowwriter commented on {{cookiecutter.project_slug}}/README.md:
```diff
@@ -1,4 +1,4 @@
-# Title: "{{cookiecutter.project_name}}"
+# {{cookiecutter.project_name}}: {{cookiecutter.project_description}}
```
the project description is typically a bit to long to go in a title. I think ti'd be best to put this on a newline below the title.

hboisgon replied:
Can! I just checked the other HydroMT plugins so we could also change the suggestion to a short title addition (HydroMT plugin for {pluginname})

aslowwriter reviewed (commented):
Couple of small remarks, but otherwise good additions. After we resolve those, it should be good to merge.

savente93 commented:
> 1. why did you use Quarto and md files for the docs rather than hydromt templates?

Couple of reasons: 
1. Quarto and md is recommended by the DSC for new projects
2. Quarto has good support for multiple languages like Julia in addition to Python which could be nice if the plugin interacts with those languages. 
3. quite a few other deltares prjects like Wflow and Rebasim already use Quarto and it seems to work quite well for them. 
4. md is a very widely used format and therefore many more tools have integrations with it, rst is not. Basically I've never seen rst used outside of python documentation 
5. On a more personal note, I really dislike Sphynx and the rst format, since they are quite error prone and cumbersome to work with and the documentation of them is not so great at times. Therefore I'd much rather encourage new projects to use more widely used/ supported tools. 

> 2. I think we should still clean the few docs there is, especially python section and decide if we include the main hydromt headers or leave it completely blank/open

I have no strong feelings on this either way. I'm happy to go with whatever you think is best. 

> 3. I added dependencies used in the template files to pyproject.toml as well as extra dependencies needed for test/dev tools. I do not know Quarto so not sure what we should add for the doc dependencies. Was I correct to add these dependencies?

You were (except for a small comment regarding mypy) Quarto doesn't need python dependencies since it's not written in python, therefore it will have to be installed by the user themselves. In the github workflows there is a seperate step to install it so there it should all be fine. 


> 4. Did you also actually try and publish one plugin and checked if linting/tests/docs are correctly passed/built?

I have back when I originally made the template. I'll do it again once your pr get's merged, but I'd rather do it while we're both happy with it as it's quite cumbersome to do.

hboisgon commented:
> > 1. why did you use Quarto and md files for the docs rather than hydromt templates?
> 
> Couple of reasons:
> 
> 1. Quarto and md is recommended by the DSC for new projects
> 2. Quarto has good support for multiple languages like Julia in addition to Python which could be nice if the plugin interacts with those languages.
> 3. quite a few other deltares prjects like Wflow and Rebasim already use Quarto and it seems to work quite well for them.
> 4. md is a very widely used format and therefore many more tools have integrations with it, rst is not. Basically I've never seen rst used outside of python documentation
> 5. On a more personal note, I really dislike Sphynx and the rst format, since they are quite error prone and cumbersome to work with and the documentation of them is not so great at times. Therefore I'd much rather encourage new projects to use more widely used/ supported tools.
> 
> > 2. I think we should still clean the few docs there is, especially python section and decide if we include the main hydromt headers or leave it completely blank/open
> 
> I have no strong feelings on this either way. I'm happy to go with whatever you think is best.
> 
> > 3. I added dependencies used in the template files to pyproject.toml as well as extra dependencies needed for test/dev tools. I do not know Quarto so not sure what we should add for the doc dependencies. Was I correct to add these dependencies?
> 
> You were (except for a small comment regarding mypy) Quarto doesn't need python dependencies since it's not written in python, therefore it will have to be installed by the user themselves. In the github workflows there is a seperate step to install it so there it should all be fine.
> 
> > 4. Did you also actually try and publish one plugin and checked if linting/tests/docs are correctly passed/built?
> 
> I have back when I originally made the template. I'll do it again once your pr get's merged, but I'd rather do it while we're both happy with it as it's quite cumbersome to do.

1. Check then I'm happy with Quarto also if the docs are generated easily without installing any extra package
2. If we use md then it means the hydromt docs template and style cannot easily be re-used right? Then maybe we dont include so much pre-prepared classic pages to fill in but keep it lean (and remove the python stuff)
3. Check
4. Nice and agreed, good to not do it too often but happy to know you did try it. So I agree for one last try when we would like to merge.

hboisgon commented:
@savente93 for the last few changes in the discussions, can you have a look and if you're happy merge and finalise this PR? I think you should have edit rights. Else I can do the couple changes.

Let's also see if we move this page to Deltares or keep it in your personal repo

savente93 commented:
> If we use md then it means the hydromt docs template and style cannot easily be re-used right? Then maybe we dont include so much pre-prepared classic pages to fill in but keep it lean (and remove the python stuff)

Not the ones from hydromt-core no, but I did include the style of either Rebasim of Wflow (Can't remember which of the two it was) so the style should still be consistent. I think it's fine like this. I'll merge this one for now, if you disagree/change your mind please open another PR and we'll remove them. 

I will ask Ali to maybe look into how we can transfer the repository.

hboisgon committed:
review changes

hboisgon committed:
minimal plugin docs

hboisgon committed:
bugfix

hboisgon commented:
Hi @savente93, I now added the few remaining changes. I removed the ``project_description`` and just replaced with ``project_tagline`` also in pyproject.toml. If you disagree, you can still merge and change in your repo.

I also added kind of a minimal and simple plugin docs to be filled in or continued. What I am not sure is how to generate python API docs with Quarto and if it possible to add easily API doc links in other pages (like in the user_guide/{{cookiecutter.model_shortname}}.qmd

Hopefully I addressed all changes. If it's the case, can you do the final check and see if all pages are produced / tests and docs works ? Many thanks :)

savente93 commented:
Looks very nice, Thanks Hélène, I think I'll merge and do the testing and editing on the repo itself. I'll make sure that everything will be ready to go there. Thanks for the contributions!
